pkg/crc/machine/start.go

```diff
@@ -318,6 +319,12 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
 		return nil, errors.Wrap(err, "Failed to change permissions to root podman socket")
 	}
 
+	if !vm.bundle.IsOpenShift() && runtime.GOOS != "darwin" {
+		if _, _, err := sshRunner.RunPrivileged("Add dns server to tap interface", "systemd-resolve --interface tap0 --set-dns 192.168.127.1"); err != nil {
+			return nil, errors.Wrap(err, "Failed to add dns server to tap interface")
```

Comment:
I spent more time on this, on the f35 coreos image, `/etc/resolv.conf` is symlinked to `/run/systemd/resolve/stub-resolv.conf`, but I don't think you are supposed to directly modify it. This is what `udhcpc.sh` ends up doing because the `gvisor-tap-vsock` container image does not have a `resolvconf` binary. Not sure how systemd-resolved is going to cope with that? (even after reading the man pages). Imo it would be best *not* to change resolv.conf from the gvisor-tap-vsock image.

Adding a systemd-resolved-friendly resolvconf binary to the image may be a bit tricky, as it's currently based on busybox in order to get a dhcp client (ubi8 does not provide one).

This PR could be a good enough workaround while we figure this out, but we need to open some follow-up issues in the appropriate places.

Reply:
On fcos-based images, we could use a gvisor-tap-vsock image using fedora as a base instead of busybox/ubi8

Reply:
> On fcos-based images, we could use a gvisor-tap-vsock image using fedora as a base instead of busybox/ubi8

@cfergeau I did trying doing it, instead of using any image I just installed dhcp client in FCOS-35 and directly run the `vm`    https://github.com/containers/gvisor-tap-vsock/blob/main/cmd/vm/main_linux.go#L132-L143 binary which should use this client but that also doesn't make use of resolved, I am not sure if resolved have a specific way to add the default nameserver provided by this virtual network dhcp server.

Reply:
For a normal interface, what happens is that NetworkManager takes care of running dhcp requests, and as part of the dhcp request, it receives DNS details. It's then NetworkManager which takes care of forwarding this information to systemd-resolved.
With gvisor-tap-vsock, `vm` takes care of running the dhcp client, not sure why. This is the code that you linked to. Then in the container image, there's a script which runs after the DHCP request and updates the DNS servers used by the system. It does it by modifying the host `/etc/resolv.conf`, which is probably not the right thing to do with `systemd-resolved`. Using `resolvconf` may work better, but it's not in our container image.
gvisor-tap-vsock seems to be sending back the DNS settings as part of DHCP, so it should just work? https://github.com/containers/gvisor-tap-vsock/blob/3ec335bea2e933a377b4da70d33fb81c63fc07f5/pkg/services/dhcp/dhcp.go#L52

Reply:
Hrm, looks like some work is needed to integrate `gvisor-tap-vsock` better with the system networking configuration:
```
$ ping gandi.net
ping: gandi.net: Name or service not known
$ sudo nmcli device modify tap0 ipv4.method auto
Connection successfully reapplied to device 'tap0'.
$ ping gandi.net
PING gandi.net (217.70.185.65) 56(84) bytes of data.
64 bytes from 217.70.185.65 (217.70.185.65): icmp_seq=1 ttl=64 time=0.245 ms
```
(I did not expect that just changing the device config would fix it right away, nor do I think that this is the 'right' fix, but this gives some starting points :)

Reworking the way gvisor-tap-vsock configures networking is probably more work than what you had in mind when filing this PR though :) I'm fine with going with a workaround for now (though I haven't looked in details at the PR yet), but the commit log needs to be more specific regarding what is broken exactly, and an issue needs to be filed so that we remember we have this to investigate more deeply some day.

Reply:
Some more experiments, creating a NM connection with `mcli conn add con-name gvisor ifname tap0 type tun mode tap` also helps (`nmcli conn up gvisor` is needed if `gvisor-tap-vsock` is already running). It's then used at subsequent boots. However, with this NM creates the tap connection automatically at boot time before `gvisor-tap-vsock` runs, and tries a dhcp request which fails. This blocks the boot process for 45s, so one more thing to fix :(

Reply:
Setting `connection.autoconnect no` helps with the boot process slowdown, but then `nmcli conn up gvisor` needs to be run manually for the connection to be used (ie for DNS to work).

Reply:
I'm not sure how to tell NM "when a tap0 device is created, then configure it using dhcp".
As a workaround, dropping this file on the image seems to be enough:
```
$ ls -al /etc/NetworkManager/dispatcher.d/pre-up.d/gvisor-tap-vsock-tap-device.sh
-rwxr-xr-x. 1 root root 143 Dec 16 10:35 /etc/NetworkManager/dispatcher.d/pre-up.d/gvisor-tap-vsock-tap-device.sh

$ cat /etc/NetworkManager/dispatcher.d/pre-up.d/gvisor-tap-vsock-tap-device.sh
#!/bin/sh

if [[ "$1" != "tap0" ]]; then
    exit 0
fi
if [[ "$2" != "pre-up" ]]; then
    exit 0
fi
nmcli device modify "$1" ipv4.method auto
```

Follow-up work would be to drop https://github.com/containers/gvisor-tap-vsock/blob/3ec335bea2e933a377b4da70d33fb81c63fc07f5/cmd/vm/main_linux.go#L132-L144 from gvisor-tap-vsock, or to replace this by the invocation of an external script, and to figure out how to integrate it better with NetworkManager.